controllers/restore_integration_test.go

```diff
@@ -0,0 +1,126 @@
+// +build integration
+
+package controllers
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	batchv1 "k8s.io/api/batch/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	k8upv1a1 "github.com/vshn/k8up/api/v1alpha1"
+)
+
+type RestoreTestSuite struct {
+	EnvTestSuite
+
+	GivenRestore *k8upv1a1.Restore
+	RestoreName  string
+}
+
+func Test_Restore(t *testing.T) {
+	suite.Run(t, new(RestoreTestSuite))
+}
+
+func (r *RestoreTestSuite) TestReconciliation() {
+	r.givenRestoreResource()
+
+	result := r.whenReconcile()
+
+	assert.GreaterOrEqual(r.T(), result.RequeueAfter, 30*time.Second)
+	r.expectAJobEventually()
+}
+
+func (r *RestoreTestSuite) SetupTest() {
+	r.RestoreName = "restore-integration-test"
+}
+
+func (r *RestoreTestSuite) TearDownTest() {
+	r.CleanUpJobs()
+	r.cleanUpRestores()
+	r.CleanUpNamespaces()
+}
```

Comment:
Please help me understand. Why do we cleanup resources, if each test within a suite is executed in a different namespace, and each suite has its own API server process that lives only as long as the tests run? Is there a reason why we add so much code to cleanup resources that we may no longer care about anyway? Namespaces provide enough isolation between tests.

We could remove 30% of the testing code.

Reply:
Now that each test is in it's own NS, this is certainly a valid argument. I like to clean up after a test, so that it's not possible to have side-effects in later test. But this is certainly not a _must_ anymore.